_posts/2020-02-09-single-responsibility-principle.md

```diff
@@ -0,0 +1,70 @@
+3. **Data**: a profile of a person on a website, a JSON document, a message. Obviously, embodying a piece of data could only be a responsibility of a class (object), but not of a method, module, or service. A specific kind of data is **configuration**: a collection of parameters for some other method, class, or system.
+4. A piece of application's **control flow, execution, or data flow** model.  One example are methods that glue together calls to other methods:
+    ```
+    void processTransaction(Transaction t) {
+      if (isFraud(t)) { // Business logic
+        // External integration: logging
+        logger.info("Detected fraud transaction {}", t);
+        // Integration with external service
+        alertingService.sendAlert(new FraudTransactionAlert(t));
+      }
+    }
+    ```
+    Another example may be a `BufferedLogger` class which buffers logging statements in memory and manages a separate background thread that takes statements from the buffer and writes them to actual external logger:
+    ```
+    class BufferedLogger implements Logger {
+      private final Logger delegate;
+      private final ExecutorService backgroundWorker;
+      private final BlockingQueue<Statement> buffer;
+
+      BufferedLogger(Logger delegate) {
+        this.delegate = delegate;
+        this.backgroundWorker = newSingleThreadExecutor();
+        this.buffer = new ArrayBlockingQueue<>(100);
+        backgroundWorker.execute(this::writeStatementsInBackground);
+      }
+      
+      @Override public void log(Statement s) {
+        putUninterruptibly(buffer, s);
+      }
+
+      private void writeStatementsInBackground() {
+        while (true) {
+          Statement s = takeUninterruptibly(buffer);
+          delegate.log(s);
+        }
+      }
+    }
+    ```
+   `BufferedLogger` is an [active object](https://en.wikipedia.org/wiki/Active_object). Method `writeStatementsInBackground` itself has control flow responsibility.
```

Comment:
You introduce the term "Active Object" without explaining it. Does it help in this context to know that `BufferedLogger` is an active object? Of not, leave this sentence away, if yes, please add a sentence explaining what it is.

Reply:
Removed sentence.